activerecord/lib/active_record/relation.rb

```diff
@@ -407,9 +407,16 @@ def scoping
       already_in_scope? ? yield : _scoping(self) { yield }
     end
 
-    def _exec_scope(name, *args, &block) # :nodoc:
+    def _exec_scope(name, *args, **kwargs, &block) # :nodoc:
```

Comment:
Can you use `ruby2_keywords(:_exec_scope) if respond_to?(:ruby2_keywords, true)`?

```diff
diff --git a/activerecord/lib/active_record/relation.rb b/activerecord/lib/active_record/relation.rb
index 4c137d0d59..55c9956012 100644
--- a/activerecord/lib/active_record/relation.rb
+++ b/activerecord/lib/active_record/relation.rb
@@ -413,6 +413,7 @@ def _exec_scope(name, *args, &block) # :nodoc:
     ensure
       @delegate_to_klass = false
     end
+    ruby2_keywords(:_exec_scope) if respond_to?(:ruby2_keywords, true)
 
     # Updates all records in the current relation with details given. This method constructs a single SQL UPDATE
     # statement and sends it straight to the database. It does not instantiate the involved models and it does not
```

Strictly speaking, `*args` to `*args, **kwargs` is not perfectly compatible except Ruby 3.0, `ruby2_keywords` is only way to do that as long as support Ruby 2.x.

Reply:
Seems like ruby2_keywords is only valid for methods taking kwargs or a splat.

```
named.rb:198: warning: Skipping set of ruby2_keywords flag for tagged_with (method accepts keywords or method does not accept argument splat)
```

I also wanted to avoid this because we _should_ see warnings in 2.7 if a named scope takes kwargs but is passed a hash.

A agree that the `*args, **kwargs` doesn't quite work here. It might be that the only way to do this is using some `if RUBY_VERSION >= ` checks and duplicating these two methods 😩

Reply:
I've added test case for `generate_relation_method` 7ce3451, it also demonstrates a regression for this PR.

Can you apply the above patch `singleton_class.send(:ruby2_keywords, name) if respond_to?(:ruby2_keywords, true)` in `lib/active_record/scoping/named.rb` instead of this PR? That patch is only way to pass 7ce3451 and address kwargs warnings in Ruby 2.7.

Reply:
Per my last comment we can't use that here as it causes even more warnings.